returns/interfaces/failable.py

```diff
@@ -0,0 +1,166 @@
+from abc import abstractmethod
+from typing import Callable, ClassVar, NoReturn, Sequence, Type, TypeVar
+
+from typing_extensions import final
+
+from returns.interfaces.bimappable import BiMappableN
+from returns.interfaces.container import ContainerN
+from returns.interfaces.rescuable import RescuableN
+from returns.primitives.asserts import assert_equal
+from returns.primitives.hkt import KindN
+from returns.primitives.laws import (
+    Law,
+    Law3,
+    Lawful,
+    LawSpecDef,
+    law_definition,
+)
+
+_FirstType = TypeVar('_FirstType')
+_SecondType = TypeVar('_SecondType')
+_ThirdType = TypeVar('_ThirdType')
+_UpdatedType = TypeVar('_UpdatedType')
+
+_DiverseFailableType = TypeVar('_DiverseFailableType', bound='DiverseFailableN')
+
+# Used in laws:
+_NewFirstType = TypeVar('_NewFirstType')
+
+
+@final
+class _FailableLawSpec(LawSpecDef):
+    """
+    Failable laws.
+
+    We need to be sure that ``.rescue`` won't rescue success types.
+    """
+
+    @law_definition
+    def rescue_short_circuit_law(
+        raw_value: _FirstType,
+        container: 'FailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_SecondType],
+            KindN['FailableN', _FirstType, _NewFirstType, _ThirdType],
+        ],
+    ) -> None:
+        """Ensures that you cannot rescue a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).rescue(function),
+        )
+
+
+@final
+class _DiverseFailableLawSpec(LawSpecDef):
```

Comment:
Looks like we should only have `alt_short_circuit_law` in `_DiverseFailableLawSpec`.
All other laws can be moved to `_FailableLawSpec`

Reply:
I think not, because `map_short_circuit_law` and `bind_short_circuit_law` use `.from_failure` method.
The `.from_failure` is only available on `DiverseFailableN`!

Reply:
Then we come back to `.empty` feature again. Because, `.map` should not work on `Nothing` or `NaN` or anything else. And we need to somehow express it.

Maybe we can create a new interfaces called `SingleFailableN`?
In this case we would have:

```
FailableN -> SingleFailableN -> Maybe
|
 \
  \
    - - - -> DiverseFailableN -> Result
```

And `SingleFailableN` can have `empty` property with `Nothing` for `Maybe`.
This way, we would be able to add laws for it as well.

What do you think?

Reply:
What are you proposing here?

Using the current scenario we'll have something like this:
```python
class FailableN(ContainerN[...], RescuableN[...]):
    ...

class SingleFailableN(FailableN[...]):
    empty: ClassVar[Any]

class DiverseFailableN(FailableN[...], BiMappableN[...]):
    ...
```

But, we won't able to move the tests to the `_FailableLawSpec` because `.from_value` have to stay on `DiverseFailableN`!

Reply:
Yes, they will have two different tests:

```python
class SingleFailableLawSpec(LawSpecDef):
    @law_definition
    def map_short_circuit_law(
        raw_value: _SecondType,
        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
        function: Callable[[_FirstType], _NewFirstType],
    ) -> None:
        """Ensures that you cannot map a failure."""
        assert_equal(
            container.empty,
            container.empty.map(function),
        )

    # etc for all other laws
```